src/libutil/fs-sink.cc

```diff
@@ -90,13 +82,18 @@ struct RestoreRegularFile : CreateRegularFileSink {
     void preallocateContents(uint64_t size) override;
 };
 
-void RestoreSink::createRegularFile(const Path & path, std::function<void(CreateRegularFileSink &)> func)
+void RestoreSink::createRegularFile(const CanonPath & path, std::function<void(CreateRegularFileSink &)> func)
 {
-    Path p = dstPath + path;
+    auto p = dstPath;
+
+    if (!path.rel().empty()) {
+        p = p / path.rel();
+    }
```

Comment:
Might be nice to factor this out into a function so we can write
```suggestion
    auto p = append(dstPath, path);
```
here and in `createSymlink()`.

Reply:
Agree, but let's do in a follow-up PR.